pil-analyzer/src/traits_resolver.rs

```diff
@@ -0,0 +1,131 @@
+use powdr_ast::{
+    analyzed::{
+        Expression, FunctionValueDefinition, Identity, PolynomialReference, Reference, Symbol,
+        TypedExpression,
+    },
+    parsed::{
+        types::{TupleType, Type},
+        visitor::AllChildren,
+        SelectedExpressions, TraitImplementation,
+    },
+};
+use std::{collections::HashMap, sync::Arc};
+
+type SolvedImpl = ((String, Vec<Type>), Arc<Expression>);
+
+pub struct TraitsResolver<'a> {
+    definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    identities: &'a Vec<Identity<SelectedExpressions<Expression>>>,
+}
+
+impl<'a> TraitsResolver<'a> {
+    pub fn new(
+        definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+        identities: &'a Vec<Identity<SelectedExpressions<Expression>>>,
+    ) -> Self {
+        Self {
+            definitions,
+            identities,
+        }
+    }
+
+    pub fn resolve_traits(
+        &'a self,
+        trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
```

Comment:
Why is this not part of the state?
Also I wonder if the code would be cleaner by moving some of the logic from this function to pil_analyzer. The trait resolver does not really need to know how to extract expressions from the various kinds of function value definitions.